modules/generator/processor/localblocks/query_range.go

```diff
@@ -101,17 +111,20 @@ func (p *Processor) QueryRange(ctx context.Context, req *tempopb.QueryRangeReque
 			continue
 		}
 
-		wg.Add(1)
-		go func(b *ingester.LocalBlock) {
-			defer wg.Done()
-			resp, err := p.queryRangeCompleteBlock(ctx, b, *req, timeOverlapCutoff, unsafe, int(req.Exemplars))
-			if err != nil {
-				jobErr.Store(err)
-				return
-			}
-
-			jobEval.ObserveSeries(resp)
-		}(b)
+		if maxSeries == 0 || (syncAtomic.LoadInt32(&totalSeriesCount) < int32(maxSeries)) {
+			wg.Add(1)
+			go func(b *ingester.LocalBlock) {
+				defer wg.Done()
+				resp, err := p.queryRangeCompleteBlock(ctx, b, *req, timeOverlapCutoff, unsafe, int(req.Exemplars))
+				if err != nil {
+					jobErr.Store(err)
+					return
+				}
+				jobEval.ObserveSeries(resp)
+				seriesCount := jobEval.Length()
+				syncAtomic.AddInt32(&totalSeriesCount, int32(seriesCount))
```

Comment:
Can you clarify this part a bit - The atomic is used to skip remaining blocks when we exceed the limit, but should it be a `Set` instead of `Add`?  It looks like if we have 100 complete blocks all returning the same 10 series, it can still hit the (default) 1000 limit.

Reply:
Ah nice catch. The atomic should just be used to keep track of the raw eval for wal and head blocks. The jobEval.length() is already tracking the series for the completed blocks. The check should now be 

raw series count + jobEval.length() > max series

Reply:
updated to use atomic bool for maxSeriesReached which will be set to true if either rawEval.length() > maxSeries or jobEval.length() > maxSeries